WIP Updates for adding the LETKF

flampouris opened the pull request:
## Description

The UMD-LETKF code is required for the hybrid portion of GODAS. With this PR, the following issues are addressed. 

1. Optimization of the workflow to run ensemble-related jobs, this is based on the code of @JianKuang-NOAA and the use of groups. In principle, running jobs for each member has simplified significantly.
 
2. Cleanup of the workflow scripts, global variables are defined at the appropriate locations, etc.

3. Add the module files for intel 18 (Default on Hera), before I lose the scripts. 

4. Add the installation of the LETKF from the https://github.com/NOAA-EMC/UMD-LETKF

This PR creates at least conflict between the copied perturbations of the first cycle and the way that the genenspert runs on the fly to generate the perturbations. This conflict does create issues and will be removed as the development with the next PR. 

### Issue(s) addressed
Partially #106, #88, #73 

- fixes #109 

## Testing

1. This update should have **NO** effect on the 3dvar case.
2. The Hofx of the ensemble is significantly faster.
3. Compilation of the code (soca, letkf, etc.) without errors 

## Dependencies
https://github.com/NOAA-EMC/UMD-LETKF

Commit:
DA_METHOD selection moved to places.yaml

Commit:
Clean the DA_METHOD

Commit:
Merge branch 'develop' into feature/letkf

Commit:
letkf ensemble structure

Commit:
Merge branch 'feature/letkf' into feature/ens_jobs

Commit:
update resource settings

Commit:
manually intake feature/letkf

Commit:
minor bug fixes

Commit:
Merge branch 'develop' into feature/ens_jobs

Commit:
bug fixes

Commit:
Merge branch 'develop' into feature/ens_jobs

Commit:
Merge branch 'develop' into feature/ens_jobs

Commit:
update workflow yamls

Commit:
update letkf workflow

Commit:
remove softlinks before "rename"

Commit:
Merge branch 'develop' into feature/ens_jobs

Commit:
Merging develop to ens_jobs

Commit:
Update the letkf_only.yaml layout to take into consideration the dependencies

Commit:
debug the feature/ens_jobs (#123)

* Processing the icec EMC files from native to ioda

* Added if statement to check if the ice_emc files exist at $PDY

* Removed JO processing from obs_prep which is already included in SetupJoCostFunction.sh
Addressing reviewer's comments.

* Hardlinks to module files have been removed and the modulefiles directory has been created (#114)

* Hardlinks to module files have been removed and the modulefiles directory is added

* The README.md is updated to reflect the changes on 1. the loaded modules and 2. address the mistake of the local directory for the soca-bundle.

* Update the letkf_only.yaml layout to take into consideration the dependencies

* Debugging the ens_job

Commit:
Adding the env file for intel18 (Default on hera) for compiling JEDI/
SOCA

Commit:
Merge branch 'feature/ens_jobs' of https://github.com/NOAA-EMC/godas into feature/ens_jobs

Commit:
Installation of UMD-LETKF

Commit:
Readme is updated and tested

Commit:
Update README.md

Commit:
merging develop to current branch

guillaumevernieres commented:
@flampouris , just tested the letkf case, seems all good. I'll run the 2 var tests and go through the code later today.

flampouris commented:
Thank you, I have more material to merge.

There is an issue with the hack of copying the perturbations for the ENVAR
at the DA_PREP. It does not break anything but the input for LETKF is not
correct. I will address this as soon as we have the model running (or with
another hack :D ).

guillaumevernieres commented on jobs/JJOB_PREP_LETKF:
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
-#set -e
+#set -x
```
Why is -e removed?

flampouris replied:
Not intentionally, probably during automerging.

flampouris replied:
I updated this and it was failing because of the soft links of ice and ocean.pert.*.nc, these soft links have been removed, and they are generated by hofx of each ensemble member.

guillaumevernieres commented on workflow/defaults/default_resources.yaml:
```diff
@@ -11,10 +11,10 @@
+  r3dvar:       [  24,    24, !timedelta "00:10:00", 12,  "3072" ]
+  obs_prep:     [  12,    12, !timedelta "00:10:00", 1,   "3072" ]
+  da_prep:      [  1,      1, !timedelta "00:30:00", 1,   "3072" ]
+  letkf_prep:   [  24,    24, !timedelta "00:10:00", 12,  "3072" ]
```
Timing change: This means we will have to change the settings when we want to do science. I understand why, but something to be aware of.
~Also, I may be wrong, but I think asking for 24 pe will translate to 2 nodes on hera, as opposed to 1 node when asking for 20 workers ... I'll check.~ Looks like I was wrong ...

flampouris replied:
- Yes, the timing change is to accommodate the conversion of the raw observations to ioda format.

- On hera, there are 40ppn, so there is no issue requesting 24p on one node.

flampouris replied:
The LETKF prep is the socahofx3d, so we can use the resources for the r3dvar. I updated this at the reviewed version of the code.

guillaumevernieres commented on workflow/defaults/default_resources.yaml:
```diff
@@ -41,8 +41,8 @@ default_resources: &default_resources
-    - batch_memory: !calc doc.godas_resource_table.r3denvar[4]
-      mpi_ranks: !calc doc.godas_resource_table.r3denvar[0]
-      walltime: !calc doc.godas_resource_table.r3denvar[2]
-      max_ppn: !calc doc.godas_resource_table.r3denvar[1]
```
Quick Q: Removed because unused?

guillaumevernieres commented:
> Thank you, I have more material to merge. There is an issue with the hack of copying the perturbations for the ENVAR at the DA_PREP. It does not break anything but the input for LETKF is not correct. I will address this as soon as we have the model running (or with another hack :D ).

No prob. I see that the letkf is not exercised yet, but fyi, I was able to build it following your readme :+1:

flampouris commented on jobs/JJOB_GEN_ENSEMBLE:
```diff
@@ -18,7 +18,4 @@ if   [ ! -d RESTART ]; then mkdir -p RESTART; fi
-
-#if   [ ! -d RESTART ]; then rm -rf RESTART; fi
-
+srun -n $npe_psoca ${SOCA_EXEC}/soca_enspert.x ./yaml/genenspert.yml 
```
The npe_psoca was remaining from the previous version, the correct is NPE, as everything else that uses the soca.

flampouris replied:
The $npe_psoca has been replaced with $NPE

flampouris commented on jobs/JJOB_PREP_SOCA:
```diff
@@ -28,7 +28,7 @@ else
     echo 'No grid found. Generating SOCA grid.'
     echo
-    srun --nodes=1 --ntasks=12 ${SOCA_EXEC}/soca_gridgen.x ./yaml/gridgen.yml
+    srun --nodes=2 --ntasks=12 ${SOCA_EXEC}/soca_gridgen.x ./yaml/gridgen.yml
```
Thats a remaining from the past, I am testing with srun -n $NPE

flampouris replied:
this is updated to srun -n $NPE

flampouris commented on workflow/defaults/da.yaml:
```diff
@@ -24,8 +24,19 @@
   DA_SLOT_LEN:  24
   DA_WNDW_OFST: 48
   DA_TS_ONLY:   0
+  DA_SLOT_LEN: 24
```
Delete the second DA_SLOT_LEN

Commit:
Addressing  comments of reviewers

Commit:
Selective removal of softlinks for running the Hofx for each member

flampouris commented:
0. It's awkward that I reply to myself. 
1. I updated the code according to the comments of @guillaumevernieres 
2. Everything that is related to SOCA has to run with the same NPE because of bump, therefore for all the soca jobs, I updated the "srun"  to be "srun -n **$NPE**" 
3. The set -e to the PREP_LETKF job forced the removal of the softlinks from the next PR to this one (see the comment in the code).

I reran the GODAS, everything is ok, please confirm.

guillaumevernieres reviewed (approved):
All good. Tested the letkf and the 3dvar but not the envar.
